### PR TITLE
Aws logstream

### PR DESCRIPTION
**Pull request by FrenchBen**
**\- What I did**
Added support for tags to the AWS log driver - The logstream flag will still over-write the tag, allowing for backwards compatibility. The tag parses the regular tag as seen in syslog, etc.

**\- How I did it**
I updated the default of logStream from container ID to use the tag parsing default. 

**\- How to verify it**
make test (although I need to add some test for tags passed as flag)
make

**\- Description for the changelog**
Update of AWS log driver to support tags. This enables support for things like `--log-opt tag="{{.ImageName}}/{{.Name}}/{{.ID}}"` as a tag for CloudWatch. 

**\- A picture of a cute animal (not mandatory but encouraged)**
Corgi

cc @samuelkarp

### CODE REVIEW

**Comment by samuelkarp** [comment]
@FrenchBen Thanks for writing this and thanks for tagging me!  At first glance this looks like https://github.com/docker/docker/pull/25633 and https://github.com/docker/docker/pull/27480.  While many of the log drivers are using `tag`, I'd still lean against it for the reason I wrote here: https://github.com/docker/docker/issues/17747#issuecomment-159110712 (tags have a different meaning in the context of AWS).

**Comment by FrenchBen** [comment]
@samuelkarp Thanks for the quick review. 
Keeping things consistent with the existing `tag` functionality was my main focus, and as you mention many drivers currently use.
Creating an aws-specific tag, such as `aws-tag` to match the current options naming, would be the go-to if it needed to have the same meaning as the AWS tags in https://github.com/docker/docker/issues/17747#issuecomment-159110712.

IMHO 
* #27480 offers too lose of a PR where the log group is dynamic (against policy since the group needs to exist for logging to happen) and ignores unknown options. 
* #25633 looks pretty good but seems to duplicate some of the template logic that currently exists in the tag. I do like the fact that it includes tests which I may re-use for this PR.

**Comment by vdemeester** [comment]
I do agree with @FrenchBen on the two other PRs. For the naming, same things, I would have liked to keep using `tag` (as other logging driver do) but I'm also okay with a aws specific one if needed — we would just have to change `ParseLogTag` to be able to *pass* which name to use.

Aside from the naming, design LGTM on this one 👼

**Review by samuelkarp — COMMENTED** [review]
@FrenchBen I'm still not wild about `tag` since the primitive here is a log stream, but I think you're right that we could create an `awslogs-tags` option if necessary.

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (vdemeester) [inline daemon/logger/awslogs/cloudwatchlogs.go]: @samuelkarp that's a good point… not sure why this (and `labels`/`env`) are not constant. But this could be done in a follow-up PR :angel:

**Comment by vdemeester** [comment]
/me tentatively add this to the 1.13.0 milestone :stuck_out_tongue_closed_eyes:

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (FrenchBen) [inline daemon/logger/awslogs/cloudwatchlogs_test.go]: Tag only over-writes the `logStreamName`; whatever restrictions were in place before, are still in place. 
Essentially I went from taking the default for `logStreamName := ctx.ContainerID` to use the default of the tag (which is also `ctx.ContainerID`), but adds the flexibility of tag parsing as seen with all other log drivers.

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (FrenchBen) [inline daemon/logger/awslogs/cloudwatchlogs.go]: Fixed!

Comment (samuelkarp) [inline daemon/logger/awslogs/cloudwatchlogs.go]: Thanks!

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (samuelkarp) [inline daemon/logger/awslogs/cloudwatchlogs_test.go]: What's shortening "container-abcdefghijklmnopqrstuvwxyz01234567890" to "container-ab"?  I would have expected this assertion to be "test-container/container-abcdefghijklmnopqrstuvwxyz01234567890".

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (FrenchBen) [inline daemon/logger/awslogs/cloudwatchlogs.go]: I don't believe this to should trigger an error - Specifying `awslogs-stream` with tag will have no error, as `awslogs-stream` is "king" and overwrites the tag specified.
 
This allows you to update the driver to use the tag specified when no logstream have been specified, but not see any change in currently running containers that have the `--log-opt awslogs-stream="someApp"` (keeps this driver backwards compatible)

Comment (samuelkarp) [inline daemon/logger/awslogs/cloudwatchlogs.go]: Is the use-case for setting `--log-opt tag="{{.Name}}/{{.ID}}"` (or similar) on the daemon so that it provides the template by default?  If so, can we get the docs updated as well?

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (FrenchBen) [inline daemon/logger/awslogs/cloudwatchlogs_test.go]: Per the docs:
https://docs.docker.com/engine/admin/logging/log_tags/
`{{.ID}}	The first 12 characters of the container id.`
`expr length "container-ab"` will output `12`

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (FrenchBen) [inline daemon/logger/awslogs/cloudwatchlogs.go]: @samuelkarp that's exactly right, the idea is to set a template on the daemon for all new containers. 
I'll make sure to create a PR for the docs to show that tag is now allowed but `awslogs-stream` would overwrite the tag.

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (samuelkarp) [inline daemon/logger/awslogs/cloudwatchlogs_test.go]: Got it.  It might be nicer to change the template used for the test to one that doesn't involve truncation (maybe `{{.FullID}}`) so it's more obvious what's going on?

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (samuelkarp) [inline daemon/logger/awslogs/cloudwatchlogs.go]: I just looked at `loggerutils.DefaultTemplate` and it's `{{.ID}}`, which would be a change in behavior from the existing driver.  Can you use `{{.FullID}}` instead to keep the existing behavior?

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (samuelkarp) [inline daemon/logger/awslogs/cloudwatchlogs_test.go]: I don't feel super strongly about this, but having to look in another location to see why it's truncated makes the test more difficult for someone unfamiliar with the codebase to read.

Comment (FrenchBen) [inline daemon/logger/awslogs/cloudwatchlogs_test.go]: More than happy to update the PR to show the `FullID` if that helps getting an LGTM from you ;)

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (samuelkarp) [inline daemon/logger/awslogs/cloudwatchlogs_test.go]: Looks like it's going to fail now because the assertion didn't get updated.

Comment (FrenchBen) [inline daemon/logger/awslogs/cloudwatchlogs_test.go]: Yup - make it fail, then show that it's working :)

**Comment by FrenchBen** [comment]
@samuelkarp  all green! 💚

**Comment by samuelkarp** [comment]
@FrenchBen Are you planning to open a docs PR after this is merged?

**Comment by FrenchBen** [comment]
@samuelkarp absolutely!

**Comment by samuelkarp** [comment]
@FrenchBen Thanks for opening the docs PR!  The only remaining feedback I have is about the default stream name when both `awslogs-stream` and `tag` are not specified (see [this comment](https://github.com/docker/docker/pull/27707#discussion_r85618848)).

**Review by vdemeester — APPROVED** [review]
design LGTM 🐸

**Comment by FrenchBen** [comment]
Missed the comment - Let me update the PR to reflect that

**Review by samuelkarp — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
@FrenchBen can you squash your commits?

**Review by vdemeester — APPROVED** [review]
LGTM 🐸 once commit squashed 👼

**Comment by FrenchBen** [comment]
:shipit:  ⚡️  @vdemeester

**Review by thaJeztah — DISMISSED** [review]
LGTM

thanks for reviewing @samuelkarp!

**Comment by cpuguy83** [comment]
@FrenchBen It'd be nice to get these into 1 commit.

**Comment by samuelkarp** [comment]
@thaJeztah Happy to help!

**Comment by thaJeztah** [comment]
oh! I thought it was squashed, darn. @FrenchBen :heart:

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
all :green: merging!

**Comment by ozbillwang** [comment]
Could anyone explain how to use this new feature with `awslogs-stream`?

could you share the `docker run` command and docker version?

I did the same as `--log-opt awslogs-stream="$HOSTNAME-{{.ID}}"`, but I didn't get the docker container id properly.

```
docker run --log-driver=awslogs --log-opt awslogs-region=ap-southeast-2
--log-opt awslogs-group="hello-world"
--log-opt awslogs-stream="$HOSTNAME-{{.ID}}"
hello-world
```
Here is the log stream name I saw, it doesn't show container id of `{{.ID}}`

```
ip-10-0-2-123.ap-southeast-2.compute.internal-{{.ID}}
```

I did try these `log_tags` one by one, seems no one works.

https://docs.docker.com/config/containers/logging/log_tags/

```
{{.ID}} | The first 12 characters of the container ID.
{{.FullID}} | The full container ID.
{{.Name}} | The container name.
{{.ImageID}} | The first 12 characters of the container’s image ID.
{{.ImageFullID}} | The container’s full image ID.
{{.ImageName}} | The name of the image used by the container.
{{.DaemonName}} | The name of the docker program (docker).
```

**Comment by samuelkarp** [comment]
Use `--log-opt tag`, not `--log-opt awslogs-stream`.  See [the documentation](https://docs.docker.com/config/containers/logging/awslogs/#tag).

**Comment by ozbillwang** [comment]
Thanks a lot, that works.
```
docker run --log-driver=awslogs --log-opt awslogs-region=ap-southeast-2
  --log-opt awslogs-create-group=true 
  --log-opt awslogs-group="hello-world"
  --log-opt tag="$HOSTNAME/{{.ID}}"
  hello-world
```

**Pull request merged**
